mm2src/crypto/src/key_derivation.rs

```diff
@@ -0,0 +1,218 @@
+    // Future algorithms can be added here.
+}
+
+/// Derives AES and HMAC keys from a given password and salts for mnemonic encryption/decryption.
+///
+/// # Returns
+/// A tuple containing the AES key and HMAC key as byte arrays, or a `MnemonicError` in case of failure.
+#[allow(dead_code)]
+pub(crate) fn derive_keys_for_mnemonic(
+    password: &str,
+    salt_aes: &SaltString,
+    salt_hmac: &SaltString,
+) -> MmResult<([u8; 32], [u8; 32]), KeyDerivationError> {
+    let argon2 = Argon2::default();
+
+    // Derive AES Key
+    let aes_password_hash = argon2.hash_password(password.as_bytes(), salt_aes)?;
+    let key_aes_output = aes_password_hash
+        .serialize()
+        .hash()
+        .ok_or_else(|| KeyDerivationError::PasswordHashingFailed("Error finding AES key hashing output".to_string()))?;
+    let key_aes = key_aes_output
+        .as_bytes()
+        .try_into()
+        .map_err(|_| KeyDerivationError::PasswordHashingFailed("Invalid AES key length".to_string()))?;
+
+    // Derive HMAC Key
+    let hmac_password_hash = argon2.hash_password(password.as_bytes(), salt_hmac)?;
+    let key_hmac_output = hmac_password_hash.serialize().hash().ok_or_else(|| {
+        KeyDerivationError::PasswordHashingFailed("Error finding HMAC key hashing output".to_string())
+    })?;
+    let key_hmac = key_hmac_output
+        .as_bytes()
+        .try_into()
+        .map_err(|_| KeyDerivationError::PasswordHashingFailed("Invalid HMAC key length".to_string()))?;
+
+    Ok((key_aes, key_hmac))
+}
+
+/// Splits a path into its components and derives a key for each component.
+fn derive_key_from_path(master_node: &[u8], path: &str) -> MmResult<[u8; 32], KeyDerivationError> {
+    let mut current_key_material = master_node.to_vec();
+    for segment in path.split('/').filter(|s| !s.is_empty()) {
+        let mut mac = HmacSha512::new_from_slice(&current_key_material[..32])
+            .map_err(|_| KeyDerivationError::HmacInitialization)?;
+        mac.update(b"\x00");
+        mac.update(segment.as_bytes());
+        drop_mutability!(mac);
+
+        let hmac_result = mac.finalize().into_bytes();
+        current_key_material = hmac_result.to_vec();
+    }
+    drop_mutability!(current_key_material);
+
+    current_key_material[32..64]
+        .try_into()
+        .map_to_mm(|_| KeyDerivationError::InvalidKeyLength)
+}
+
+/// Derives encryption and authentication keys from the master private key using [SLIP-0021](https://github.com/satoshilabs/slips/blob/master/slip-0021.md).
+///
+/// # Returns
+/// A tuple containing the encryption and authentication keys as byte arrays, or a [`KeyDerivationError`] in case of failure.
+#[allow(dead_code)]
+pub(crate) fn derive_encryption_authentication_keys(
+    master_secret: &[u8; 64],
+    encryption_path: &str,
+    authentication_path: &str,
+) -> MmResult<([u8; 32], [u8; 32]), KeyDerivationError> {
+    const SYMMETRIC_KEY_SEED: &[u8] = b"Symmetric key seed";
```

Comment:
Is this hard-coded seed part of the actual algorithm?

Reply:
Yes, https://github.com/satoshilabs/slips/blob/master/slip-0021.md#master-node-generation

Reply:
I think a better const name is `HMAC_KEY_FOR_MASTER_NODE_DERIVATION` or just `MASTER_NODE_HMAC_KEY` will change it.

Reply:
Done